API for role assignments

davidfowl opened the pull request:
This adds support for APIs for modeling a more secure default security configuration when deploying to azure container apps. It includes a few things:

1. A new set of APIs that azure resources to expose a way to define per reference role assignments. This allows for a least privilege access to resources:

```C#
using Azure.Provisioning.Storage;

var builder = DistributedApplication.CreateBuilder(args);

var storage = builder.AddAzureStorage("storage")
                   .RunAsEmulator(c => c.WithLifetime(ContainerLifetime.Persistent));
var blobs = storage.AddBlobs("blobs");

builder.AddProject<Projects.AzureContainerApps_ApiService>("api")
       .WithExternalHttpEndpoints()
       .WithReference(blobs)
       .WithRoleAssignments(storage, StorageBuiltInRole.StorageBlobDataContributor);

builder.Build().Run();
```

In the above `WithRoleAssignments` allows the user to specify which roles this project/container requires what roles for a specific storage account.

2. Each compute resource that becomes a container app, gets a roles bicep resource that allocates the managed identity and sets up specific role assignments for that identity based on both the either the per reference role assignments or the resource defaults.
    a. If the targeted resource is an existing resource, it needs to go into its own bicep resource so we can set the "Scope" correctly.

3. Role assignment scopes are resource references. This means we need the ability to get a reference to the underlying azure resource by name, so we can use an existing bicep resource reference in "roles" bicep module. This led to a strange API that was added to `AzureProvisioningResource` to do this in a way that avoids the compute resource needing to know how to get a reference to the underlying azure resource.
 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/6636)

davidfowl committed:
Spike of API

davidfowl committed:
Next step

davidfowl committed:
Next phase

davidfowl committed:
Least privilege

davidfowl committed:
More tweaks

davidfowl committed:
Cleaned up some APIs

davidfowl commented on playground/AzureContainerApps/AzureContainerApps.AppHost/api-roles.module.bicep:
```diff
@@ -0,0 +1,27 @@
+
+param storage_outputs_name string
+
+resource identity 'Microsoft.ManagedIdentity/userAssignedIdentities@2023-01-31' = {
```
We generate a module per container/project and that module has both the user assigned identity and role assignments for that compute resource.

davidfowl commented on playground/AzureContainerApps/AzureContainerApps.AppHost/api-roles.module.bicep:
```diff
@@ -0,0 +1,27 @@
+  scope: storage
+}
+
+output id string = identity.id
+
+output clientId string = identity.properties.clientId
```
These end up as outputs for the container app to use for both runtime and to associate this user assigned identity with the container app.

davidfowl commented on playground/AzureContainerApps/AzureContainerApps.AppHost/api-roles.module.bicep:
```diff
@@ -0,0 +1,27 @@
+  location: location
+}
+
+resource storage 'Microsoft.Storage/storageAccounts@2024-01-01' existing = {
+  name: storage_outputs_name
+}
```
We need to reference the set of resources we are creating role assignments for.

davidfowl committed:
Fixed tests and clean up processing
- Added descriptions to role assignments
- Fixed tests with new output

davidfowl committed:
Name the identity after the compute resource to avoid conflicts

davidfowl committed:
Added RoleDefinition

davidfowl commented on src/Aspire.Hosting.Azure/RoleDefinition.cs:
```diff
@@ -0,0 +1,13 @@
+/// <param name="Description"></param>
+
+// REVIEW: This should be part of the Azure.Provisioning APIs
+public record struct RoleDefinition(string Id, string Description);
```
I think we need a single type for role definitions and then per resource types that have the well known ones. That would reduce the API explosion.

cc @tg-msft

davidfowl committed:
Fix tests

davidfowl commented on src/Aspire.Hosting.Azure.Storage/AzureStorageResource.cs:
```diff
@@ -83,4 +86,16 @@ void IResourceWithAzureFunctionsConfig.ApplyAzureFunctionsConfiguration(IDiction
+    /// <returns></returns>
+    public override ProvisionableResource GetExistingResource(BicepValue<string> name)
+    {
+        var account = StorageAccount.FromExisting(Infrastructure.NormalizeBicepIdentifier(Name));
+        account.Name = name;
```
This the strangest part of this change. We need the ability to import the underlying resource use for the role assignment given a name. Look at [this](https://github.com/dotnet/aspire/pull/6636/files#diff-5c542bf0932589f40e044926646e54c8fed701df0595fcdf18025abbe80aaea3R11) to understand more.

This method is called from [here](https://github.com/dotnet/aspire/pull/6636/files#diff-40fca6106cdf3968888c9d2bf55bb47a99dd1087a005c291248d2783dcd5dc07R330)

davidfowl committed:
Changed the order

mitchdenny commented:
Have you tried deploying this? Based on the Bicep being generated it looks like you are deploying the `-roles.module.bicep` file first and then taking the user assigned identity out of that. I think that is fine for the case where container apps are only accessing other resources - but we just need to double check this model work for container app to container app deployments.

I think it'll be OK because for container app to container app communication the container app identity (resource side) is not what matters, there would need to be a different Azure Entra app registration which represents the resource (which we are going to need to support defining in the app model).

You know if we get this working it's going to take a lot of the friction out of doing S2S auth.

davidfowl commented:
> Have you tried deploying this? Based on the Bicep being generated it looks like you are deploying the -roles.module.bicep file first and then taking the user assigned identity out of that. I think that is fine for the case where container apps are only accessing other resources - but we just need to double check this model work for container app to container app deployments.

Not yet, there was a bug in azd, [fixed in this pr](https://github.com/Azure/azure-dev/pull/4524). It'll work because the identity and role assignments are part of the provision phase. I'll update the description with the deployment details.

davidfowl commented:
Deployment works!

eerhardt committed:
Merge remote-tracking branch 'upstream/main' into davidfowl/role-assignments

eerhardt committed:
Fix merge. Get build passing. Get tests passing. Update playground.

eerhardt commented on src/Aspire.Hosting.Azure/AzureProvisioningResourceExtensions.cs:
```diff
@@ -198,5 +198,16 @@ private static ProvisioningParameter GetOrAddParameter(AzureResourceInfrastructu
         return parameter;
     }
+
+    private static string GetNameFromValueExpression(IManifestExpressionProvider ep)
```
@davidfowl - do you remember why you made this change? Is it still necessary?

eerhardt committed:
Support existing resources

eerhardt committed:
Add a Service when the publisher supports role assignments so the resources know not to do it themselves.

eerhardt committed:
- Move WithRoleAssignments to the parent storage resource.
- Use AzureProvisioningOptions to tell whether specific role assignments can be used, or if the default role assignments should be used.
- Throw if specific role assignments are specified but won't be honored by the publisher.

eerhardt committed:
Merge remote-tracking branch 'upstream/main' into davidfowl/role-assignments

eerhardt committed:
Move role assignment processing logic into a new lifecycle hook called AzureResourcePreparer.

Introduce a new "AppliedRoleAssignmentsAnnotation"
 - This is applied to an Azure resource to tell it which RoleAssignments it should generate in its own bicep template for the KnownParameters PrincipalId and PrincipleType
 - Used when not using ACA infra
   - This is both for
     - azd up's infra
     - F5 dev time provisioning
 - When using ACA infra during publish, AppliedRoleAssignments will be empty

When not using ACA infra, just copy all the DefaultRoleAssignments to AppliedRoleAssignments

When using ACA infra, walk the resource graph and
- if in RunMode:
  - If a compute resource has RoleAssignmentAnnotations, add them to AppliedRoleAssignmentsAnnotation on the referenced Azure resource
  - if the resource doesn't, copy the DefaultRoleAssignments to AppliedRoleAssignmentsAnnotation

- if in PublishMode:
  - If a compute resource has RoleAssignmentAnnotations, skip - the ACA publish infrastructure will handle them
  - if the compute resource doesn't have targeted RoleAssignments, copy the DefaultRoleAssignments to RoleAssignmentAnnotations so the publish infrastucture will apply the defaults

eerhardt commented:
@davidfowl @mitchdenny @sebastienros - let me know what you think about the latest iteration. It feels like an OK solution, but I'm sure there are improvements that can be made.

eerhardt committed:
Fix tests

eerhardt committed:
Clean up and docs

eerhardt committed:
More cleanup

eerhardt commented:
Taking this out of draft, as the code changes are ready for review. I still need to write more tests before I want to merge this. But the code changes should all be there.

eerhardt committed:
Write more unit tests.
Clean up code in preparation for review.

eerhardt committed:
Merge remote-tracking branch 'upstream/main' into davidfowl/role-assignments

davidfowl commented on src/Aspire.Hosting.Azure/AzureProvisioningOptions.cs:
```diff
@@ -19,4 +19,10 @@ public sealed class AzureProvisioningOptions
+    /// Gets or sets a value indicating whether Azure resources should include default role assignments
+    /// in their bicep templates.
+    /// </summary>
+    public bool UseDefaultRoleAssignments { get; set; } = true;
```
Should we flip this to make default bool be false?

davidfowl commented on src/Aspire.Hosting.Azure/AzureResourcePreparer.cs:
```diff
@@ -0,0 +1,248 @@
+
+    private async Task<HashSet<IAzureResource>> GetAzureReferences(IResource resource, CancellationToken cancellationToken)
+    {
+        if (resource.TryGetEnvironmentVariables(out var environmentCallbacks))
```
You need to do arguments as well.

davidfowl commented on tests/Aspire.Hosting.Azure.Tests/AzureManifestUtils.cs:
```diff
@@ -0,0 +1,41 @@
+        {
+            var executionContext = new DistributedApplicationExecutionContext(DistributedApplicationOperation.Publish);
+            var azurePreparer = new AzureResourcePreparer(Options.Create(new AzureProvisioningOptions()), executionContext);
+            var resources = (IResourceCollection)Activator.CreateInstance(
+                typeof(IResourceCollection).Assembly.GetType("Aspire.Hosting.ApplicationModel.ResourceCollection")!)!;
+            resources.Add(resource);
+            await azurePreparer.BeforeStartAsync(new DistributedApplicationModel(resources), cancellationToken: default);
```
What is this jank?

eerhardt replied:
Which part do you hate?
* Having to call AzurePreparer before generating the bicep?
* Reusing the internal ResourceCollection class via Reflection? If this is an issue, I can create a copied implementation of IResourceCollection.

eerhardt replied:
I've updated the code to no longer use reflection. I added a convienence ctor on DistributedApplicationModel. Let me know if you think this is acceptable.

eerhardt committed:
PR feedback

eerhardt committed:
Make it easier to create a DistributedApplicationModel instance.

eerhardt reviewed (approved):
LGTM. Approving because @davidfowl opened this PR and can't approve his own PR.
